pre_commit/languages/download.py

```diff
@@ -0,0 +1,267 @@
+
+        self.__hasher = hashlib.new(self.algorithm)
+
+        if len(b64decode(self.checksum)) != self.__hasher.digest_size:
+            raise ValueError(
+                f'Invalid checksum string length for {self.algorithm}',
+            )
+
+    @property
+    def value(self) -> str:
+        return self.__value
+
+    @property
+    def algorithm(self) -> str:
+        return self._algorithm
+
+    @property
+    def checksum(self) -> str:
+        return self._checksum
+
+    @property
+    def digest(self) -> str:
+        return b64encode(self.__hasher.digest()).decode('utf-8')
+
+    def update(self, data: bytes) -> None:
+        return self.__hasher.update(data)
+
+    def check_content(self) -> None:
+        if self.digest != self.checksum:
+            raise ChecksumError(self)
+
+    def __str__(self) -> str:
+        return self.value
+
+
+@dataclass(frozen=True)
+class URI:
+    value: str
+
+    def __post_init__(self) -> None:
+        url = urlparse(self.value)
+        if not all([url.scheme, url.netloc]):
+            raise ValueError(f'Invalid URI: {self.value}')
+
+    def __str__(self) -> str:
+        return self.value
+
+
+@dataclass(frozen=True)
+class Metadata:
+    value: str
+
+    @property
+    def parts(self) -> tuple[str, str, str, str]:
+        first, second, third, fourth = self.value.splitlines()
+        return (first, second, third, fourth)
+
+    @property
+    def platform(self) -> Platform:
+        platform, _, _, _ = self.parts
+        return Platform(platform)
+
+    @property
+    def sri(self) -> SRI:
+        _, sri, _, _ = self.parts
+        return SRI(sri)
+
+    @property
+    def uri(self) -> URI:
+        _, _, uri, _ = self.parts
+        return URI(uri)
+
+    @property
+    def filename(self) -> PurePath:
+        _, _, _, path = self.parts
+        return PurePath(path)
+
+
+def download(uri: URI, sri: SRI, filename: Path) -> None:
+    filename.parent.mkdir(parents=True, exist_ok=True)
+    with urlopen(str(uri)) as ws:
+        with filename.open('wb') as fp:
+            while buf := ws.read(IO_BUF_SIZE):
+                fp.write(buf)
+                sri.update(buf)
+            fp.flush()
+    sri.check_content()
+    chmod(filename, stat.S_IRUSR | stat.S_IXUSR)
+
+
+def install_environment(
+        prefix: Prefix,
+        version: str,
+        additional_dependencies: Sequence[str],
+) -> None:
+    host = Platform.host()
+    for dep in additional_dependencies:
+        m = Metadata(dep)
+        if host == m.platform:
+            envdir = Path(
+                lang_base.environment_dir(
+                    prefix, ENVIRONMENT_DIR,
+                    version,
+                ),
+            )
+            filename = envdir / m.filename
+            download(m.uri, m.sri, filename)
+            srisum = envdir / 'health.srisum'
+            with srisum.open('w', encoding='utf8') as stream:
+                stream.write(f'{m.sri}  {filename}\n')
```

Comment:
Files are relative to the current sumfile:

```suggestion
                stream.write(f'{m.sri}  {filename.relative_to(srisum)}\n')
```

Reply:
```
fuffan01@e125767:/tmp/pytest-of-fuffan01/pytest-11$ cat test_download_healthy0/prefix/download-default/health.srisum
sha256-oRJkj6Cr8nWIivZ9d3W+rVZt/aSW1l9YtxSVh+GtIHM=  test.bat
```
This is what the ``srisum`` file is showing. I think this comment is resolved. Feel free to re-open it.